x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,436 @@
+            out.writeArray(StreamOutput::writeBytesRef, key);
+        }
+
+        public String asString() {
+            StringBuilder builder = new StringBuilder();
+            for (int i = 0; i < key.length - 1; i++) {
+                builder.append(key[i].utf8ToString()).append(" ");
+            }
+            builder.append(key[key.length - 1].utf8ToString());
+            return builder.toString();
+        }
+
+        @Override
+        public String toString() {
+            return asString();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            BucketKey bucketKey = (BucketKey) o;
+            return Arrays.equals(key, bucketKey.key);
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(key);
+        }
+
+        public BytesRef[] keyAsTokens() {
+            return key;
+        }
+
+        @Override
+        public int compareTo(BucketKey o) {
+            return Arrays.compare(key, o.key);
+        }
+
+    }
+
+    public static class Bucket extends InternalBucket implements MultiBucketsAggregation.Bucket, Comparable<Bucket> {
+        // Used on the shard level to keep track of sub aggregations
+        long bucketOrd;
+
+        final BucketKey key;
+        final long docCount;
+        InternalAggregations aggregations;
+
+        public Bucket(BucketKey key, long docCount, InternalAggregations aggregations) {
+            this.key = key;
+            this.docCount = docCount;
+            this.aggregations = aggregations;
+        }
+
+        public Bucket(StreamInput in) throws IOException {
+            key = new BucketKey(in);
+            docCount = in.readVLong();
+            aggregations = InternalAggregations.readFrom(in);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(CommonFields.DOC_COUNT.getPreferredName(), docCount);
+            builder.field(CommonFields.KEY.getPreferredName());
+            key.toXContent(builder, params);
+            aggregations.toXContentInternal(builder, params);
+            builder.endObject();
+            return builder;
+        }
+
+        BucketKey getRawKey()  {
+            return key;
+        }
+
+        @Override
+        public Object getKey() {
+            return key;
+        }
+
+        @Override
+        public String getKeyAsString() {
+            return key.asString();
+        }
+
+        @Override
+        public long getDocCount() {
+            return docCount;
+        }
+
+        @Override
+        public Aggregations getAggregations() {
+            return aggregations;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
```

Comment:
Could you move this method so it's right under the ctor that does the reading? That way its easier to look at them both on the same page and see that they are mirror images.